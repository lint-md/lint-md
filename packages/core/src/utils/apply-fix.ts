import { Fix } from '../types';

/**
 * 基于多个 fix 来修复一个字符串，关于 fix 的数据结构请查看相关类型定义
 *
 * @date 2021-12-14 15:48:27
 */
export const applyFix = (content: string, fixes: Fix[]) => {
  // 对所有的 fix 进行排序
  fixes.sort((a, b) => {
    return a.range[0] - b.range[0] || a.range[1] - b.range[1];
  });

  // 初始化数据
  let result = '';
  let currentIndex = Number.NEGATIVE_INFINITY;

  // 未被处理的 fix
  const notAppliedFixes: Fix[] = [];

  const tryApplyOneFix = (fix: Fix) => {
    const [start, end] = fix.range;

    // 不合法 range
    if (start > end) {
      return;
    }

    // 如果新的 fix 的起始点在当前节点的前面（这是多个 fix 重合导致的，我们会跳过它，并在下一次 fix 中尝试去 fix 它）
    if (currentIndex >= start) {
      notAppliedFixes.push(fix);
      return;
    }

    // 追加 fix range 之前的内容
    result += content.slice(currentIndex, Math.max(0, start));
    // 追加 fix 新增的文本
    result += fix.text;
    // 将当前索引指向 fix range 的末尾
    currentIndex = end;
    return;
  };

  for (const fix of fixes) {
    tryApplyOneFix(fix);
  }

  // 不要忘记追加末尾的内容
  result += content.slice(Math.max(0, currentIndex));

  return {
    result: result,
    notAppliedFixes: notAppliedFixes
  };
};
