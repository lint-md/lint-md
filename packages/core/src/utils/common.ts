/**
 * 判断是否为一个合法的 ast 节点
 *
 * @date 2021-12-12 22:01:31
 */
export const isNode = (x: any) => {
  return x !== null && typeof x === 'object' && typeof x.type === 'string';
};
