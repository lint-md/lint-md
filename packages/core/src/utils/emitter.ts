/**
 * 初始化一个发布订阅监听器
 *
 * @date 2021-12-05 15:50:56
 */
export const createEmitter = () => {
  const listeners: Record<string, any> = {};

  const on = (eventName, listener) => {
    if (listeners[eventName]) {
      listeners[eventName].push(listener);
    } else {
      listeners[eventName] = [listener];
    }
  };

  const emit = (eventName, ...args: any[]) => {
    if (listeners[eventName]) {
      listeners[eventName].forEach(listener => listener(...args));
    }
  };

  const getEventNames = () => {
    return Object.keys(listeners);
  };

  return {
    on: on,
    emit: emit,
    getEventNames: getEventNames
  };
};
