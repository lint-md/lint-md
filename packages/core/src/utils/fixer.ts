import { TextRange } from '../types';

/**
 * 初始化一个 fixer (修复器）
 * fixer 有很多 fix 方法，每一个 fix 方法被调用则会对一个字符串进行 fix
 * 其中每一次 fix 有以下参数：
 * - range 被替换的文本区间
 * - text 新增的文本
 */
export const createFixer = () => {
  /**
   * 将文本插入源文本中指定索引的位置
   *
   * @date 2021-12-14 14:19:40
   * @param index 索引的开始点
   * @param text 文本内容
   */
  const insertTextAt = (index: number, text: string) => {
    return {
      range: [index, index],
      text: text
    };
  };

  /**
   * 从某个范围之后添加文本
   *
   * @date 2021-12-14 14:29:02
   * @param range 文本范围
   * @param text 文本内容
   */
  const insertTextAfterRange = (range: TextRange, text: string) => {
    return insertTextAt(range[1], text);
  };

  /**
   * 从某个范围之前添加文本
   *
   * @date 2021-12-14 14:29:02
   * @param range 文本范围
   * @param text 文本内容
   */
  const insertTextBeforeRange = (range: TextRange, text: string) => {
    return insertTextAt(range[0], text);
  };


  /**
   * 替换某个范围的文本
   *
   * @date 2021-12-14 14:29:02
   * @param range 文本范围
   * @param text 文本内容
   */
  const replaceTextRange = (range: TextRange, text: string) => {
    return {
      range: range,
      text: text
    };
  };

  /**
   * 移除某个范围的文本
   *
   * @date 2021-12-14 14:29:02
   * @param range 文本范围
   */
  const removeRange = (range: TextRange) => {
    return {
      range: range,
      text: ''
    };
  };


  return {
    insertTextAt,
    insertTextAfterRange,
    insertTextBeforeRange,
    replaceTextRange,
    removeRange
  };
};
