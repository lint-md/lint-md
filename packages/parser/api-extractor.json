{
  "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
  "mainEntryPointFilePath": "<projectFolder>/lib/tmp/index.d.ts",
  "apiReport": {
    "enabled": false
  },
  "docModel": {
    "enabled": false
  },
  "dtsRollup": {
    "enabled": true,
    "untrimmedFilePath": "<projectFolder>/lib/lint-md-parser.d.ts"
  },
  "messages": {
    "compilerMessageReporting": {
      "default": {
        "logLevel": "warning"
      }
    },
    "extractorMessageReporting": {
      "default": {
        "logLevel": "warning"
      }
    },
    "tsdocMessageReporting": {
      "default": {
        "logLevel": "warning"
      }
    }
  }
}
