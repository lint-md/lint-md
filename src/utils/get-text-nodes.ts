/**
 * 给定一个 Markdown 节点，递归寻找所有文本节点
 * @author YuZhanglong <loveyzl1123@gmail.com>
 */
import type { MarkdownNode } from '@lint-md/parser';

type MarkdownTextNode = MarkdownNode & {
  value?: string
};

export const getTextNodes = (node: MarkdownTextNode) => {
  const textNodes: MarkdownTextNode[] = [];
  const textOnlyNode = ['text', 'inlineCode'];

  // 有 value 字段被认为是 textNode，结束递归
  if (textOnlyNode.includes(node.type)) {
    textNodes.push(node);
    return textNodes;
  }

  if (node.children) {
    const childNodes = node.children.reduce((prev, current) => {
      return prev.concat(getTextNodes(current as MarkdownTextNode));
    }, []);
    textNodes.push(...childNodes);
  }

  return textNodes;
};
