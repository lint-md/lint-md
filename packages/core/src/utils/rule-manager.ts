import { isFunction } from 'lodash';
import { ReportOption } from '../types';
import { createFixer } from './fixer';

/**
 * 初始化全局 rule 管理器
 *
 * @date 2021-12-24 22:57:11
 */
export const createRuleManager = () => {
  // 修复器
  const fixer = createFixer();

  // 已经上报的数据
  const allReportedData: ReportOption[] = [];

  // 上报方法，供选择器内部调用
  const report = (option: ReportOption) => {
    allReportedData.push(option);
  };

  // 获取所有上报的数据
  const getReportData = () => {
    return allReportedData;
  };

  // 获取所有的 fix
  const getAllFixes = () => {
    return allReportedData
      .filter(item => isFunction(item.fix))
      .map(item => item.fix(fixer));
  };

  // 初始化一个 rule context
  const createRuleContext = (options?: Record<string, any>) => {
    return {
      report: report,
      options: options || {}
    };
  };

  return {
    getReportData: getReportData,
    getAllFixes: getAllFixes,
    createRuleContext: createRuleContext
  };
};
