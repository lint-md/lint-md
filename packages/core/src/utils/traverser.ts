import { noop } from 'lodash';
import { MarkdownNode } from '@lint-md/parser';
import { TraverserOptions } from '../types';
import { isNode } from './common';

/**
 * 初始化遍历器
 *
 * @date 2021-12-12 22:04:25
 */
export const createTraverser = (options: TraverserOptions) => {
  const { onLeave = noop, onEnter = noop } = options;

  const traverse = (node: MarkdownNode, parent: MarkdownNode) => {
    if (!isNode(node)) {
      return;
    }

    onEnter(node, parent);

    const children = node?.children || [];

    // 递归处理各个子节点
    for (const child of children) {
      traverse(child, node);
    }

    onLeave(node, parent);
  };

  return {
    traverse: traverse,
  };
};
