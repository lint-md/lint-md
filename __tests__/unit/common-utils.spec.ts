import { parseMd } from '@lint-md/parser';
import { getTextNodes } from '../../src/utils/get-text-nodes';

describe('test common utils', () => {
  test('test get text node', () => {
    const res = getTextNodes(parseMd('这就是 ~~删除线~~ ![12312313](213213) **啦啦啦**<div>~~123123~~!<a>测试测试</a></div> [123123123!!](12312313)'));
    expect(res).toMatchSnapshot();
  });
});
