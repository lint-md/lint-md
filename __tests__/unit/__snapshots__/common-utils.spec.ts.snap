// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test common utils test get text node 1`] = `
Array [
  Object {
    "position": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": "text",
    "value": "这就是 ",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "start": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": "text",
    "value": "删除线",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": "text",
    "value": " ",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 33,
        "line": 1,
        "offset": 32,
      },
      "start": Object {
        "column": 32,
        "line": 1,
        "offset": 31,
      },
    },
    "type": "text",
    "value": " ",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 38,
        "line": 1,
        "offset": 37,
      },
      "start": Object {
        "column": 35,
        "line": 1,
        "offset": 34,
      },
    },
    "type": "text",
    "value": "啦啦啦",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 53,
        "line": 1,
        "offset": 52,
      },
      "start": Object {
        "column": 47,
        "line": 1,
        "offset": 46,
      },
    },
    "type": "text",
    "value": "123123",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 56,
        "line": 1,
        "offset": 55,
      },
      "start": Object {
        "column": 55,
        "line": 1,
        "offset": 54,
      },
    },
    "type": "text",
    "value": "!",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 63,
        "line": 1,
        "offset": 62,
      },
      "start": Object {
        "column": 59,
        "line": 1,
        "offset": 58,
      },
    },
    "type": "text",
    "value": "测试测试",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 74,
        "line": 1,
        "offset": 73,
      },
      "start": Object {
        "column": 73,
        "line": 1,
        "offset": 72,
      },
    },
    "type": "text",
    "value": " ",
  },
  Object {
    "position": Object {
      "end": Object {
        "column": 86,
        "line": 1,
        "offset": 85,
      },
      "start": Object {
        "column": 75,
        "line": 1,
        "offset": 74,
      },
    },
    "type": "text",
    "value": "123123123!!",
  },
]
`;
